File: clients/ruby/ip2cc.rb
class Ip2cc
  
  VALID_IP = Regexp.new('^'+(['(?:(?:2[0-4]|1\d|[1-9])?\d|25[0-5])']*4).join('\.')+'$')
  
  def initialize(filename)
    @filename = filename
  end
  
  def [](ip)
    return nil unless ip =~ VALID_IP
    parts = ip.split('.').map {|part| part.to_i}
    open(@filename) do |f|
      offset = 0
      parts.each do |part|
       start = offset + part * 4
       f.seek(start)
       value = f.read(4)
       if value[0, 2] == "\xFF\xFF" then
         return nil if value[2, 2] == "\x00\x00"
         return value[2, 2]
       end
       offset = value.unpack('N')[0]
      end
    end
    raise 'broken db'
  end
  
end

if __FILE__ == $0 then
  db = Ip2cc.new '../../ip2cc.db'
  if ARGV.length == 1 then
    puts(db[ARGV[0]])
  else
    ARGV.each{ |ip| puts "#{ip} #{db[ip]}" }
  end
end

File: clients/php/ip2cc.php
<?php

class IP2CC implements ArrayAccess {
  
  public $filename;
  
  public function __construct($filename) {
    $this->filename = $filename;
  }
  
  public function offsetGet($ip) {
    $fp = fopen($this->filename, 'rb');
    $parts = explode('.', $ip);
    $offset = 0;
    foreach($parts as $part) {
      $start = $offset + intval($part) * 4;
      fseek($fp, $start);
      $value = fread($fp, 4);
      if (substr($value, 0, 2) == "\xFF\xFF") {
        $res = substr($value, 2, 2);
        if ($res == "\x00\x00") {
          fclose($fp);
          return;
        }
        fclose($fp);
        return $res;
      }
      $chunks = unpack('N', $value);
      $offset = $chunks[1];
    }
    fclose($fp);
    throw new Exception('broken db');
  }
  
  public function offsetExists($foo) {
  }
  
  public function offsetUnset($bar) {
  }
  
  public function offsetSet($spam, $eggs) {
  }
  
}

if ($_SERVER['argc'] > 1 && realpath($_SERVER['argv'][0]) == __file__ ){
  
  $a = new IP2CC('../../ip2cc.db');
  $ip = $_SERVER['argv'][1];
  print "{$a[$ip]}\n";
  
}